src/pad.js

```
@@ -0,0 +1,111 @@
+    for (let i = 0; i < rank; i++) {
+      if (mappedLocation[i] < beginningPadding[i]) {
+        // Full equation:
+        // mappedLocation[i] =
+        //    beginningPadding[i] + (beginningPadding[i] - mappedLocation[i]) -
+        //    beginningPadding[i] - offset;
+        mappedLocation[i] = beginningPadding[i] - mappedLocation[i] - offset;
+      } else if (mappedLocation[i] >= beginningPadding[i] + inputShape[i]) {
+        // Full equation:
+        // mappedLocation[i] =
+        //    beginningPadding[i] + inputShape[i] - 1 -
+        //    (mappedLocation[i] - (beginningPadding[i] + inputShape[i] -1)) -
+        //    beginningPadding[i] + offset;
+        mappedLocation[i] =
+            beginningPadding[i] + inputShape[i] * 2 - mappedLocation[i] - 2 + offset;
+      } else {
+        mappedLocation[i] -= beginningPadding[i];
+      }
+    }
+  }
+  return mappedLocation;
+}
+
+/**
+ * Update element value of destination tensor.
+ * @param {Number} index
+ * @param {Tensor} source
+ * @param {Tensor} destination
+ * @param {Array} beginningPadding
+ * @param {MLPaddingMode} mode
+ * @param {Number} value
+ * @return {Tensor}
+ */
+function updateOutputElement(index, source, destination, beginningPadding, mode, value) {
+  const sourceShape = source.shape;
+  const location = destination.locationFromIndex(index);
+  const rank = location.length;
+  let hasFillPadding = false;
+  for (let j = 0; j < rank; j++) {
+    if (location[j] < beginningPadding[j] || location[j] >= beginningPadding[j] + sourceShape[j]) {
+      hasFillPadding = true;
+      break;
+    }
+  }
+  let result;
+  if (hasFillPadding) {
+    if (mode === 'constant') {
+      result = value;
+    } else if (mode === 'edge' || mode === 'reflection' || mode === 'symmetric') {
+      const targetLocation = getMappedLocation(location, sourceShape, beginningPadding, mode);
+      result = source.getValueByLocation(targetLocation);
+    }
```

should report an error if there are any other (invalid) modes?